_posts/2015-11-27-simple-rust-part-three.markdown

```diff
@@ -0,0 +1,374 @@
+---
+layout: post
+title: "Simple Rust, Part Three"
+date: 2015-11-27T13:37:28-06:00
+subtitle: "Vectors, Mutations, Macros and Lifetimes"
+author: Ian Whitney
+---
+
+[Previously](http://designisrefactoring.com/2015/11/13/simple-rust-part-two/), we looked at how our function will use Arrays and whether or not it will own or borrow its inputs. Our code currently looks like this:
+
+{% highlight rust %}
+fn anagrams_for(s: &str, inputs: &[&str]) {
+  //Amazing logic TBD
+}
+{% endhighlight %}
+
+We're still trying to get tests like these to pass:
+
+{% highlight rust %}
+fn test_detect_simple_anagram() {
+  let inputs = ["tan", "stand", "at"];
+  let outputs: Vec<&str> = vec!["tan"];
+  assert_eq!(anagram::anagrams_for("ant", &inputs), outputs);
+}
+{% endhighlight %}
+
+The test asserts that `outputs` is equal to what our `anagrams_for` function returns. But our function currently returns nothing. And while the `ouptputs` syntax has a few familiar aspects, there's a lot of new code as well: `Vec`, `vec!` and those angle brackets in `<&str>`. What's going on here?
+
+<!--break-->
+
+### Vectors, the mutable array
+
+Let's start with `Vec`. In the [last post](http://designisrefactoring.com/2015/11/13/simple-rust-part-two/) we looked at Array, which suited our needs because it is immutable. I mentioned that there were also mutable arrays, and here they are -- Vectors, or Vec for short.
+
+Remember that when we write an array like this:
+
+{% highlight rust %}
+let an_array = ["tan", "stand", "at"];
+{% endhighlight %}
+
+it doesn't have the type of `Array`. Instead its type is `[&str,3]`, the combination of the type and number of its elements.
```

Comment:
Array types use `;` to separate the type of items, and the length: The type should be `[&str;3]`, or, more pedantically, `[&'static str; 3]`